### ISSUE TITLE
CUDA IPC with managed memory

### ISSUE THREAD

**Issue by pentschev** [rapidsai/ucx-py#409]
CUDA IPC doesn't support managed memory, as documented [here](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#interprocess-communication). This is a problem for some cases, such as when we have larger-than-memory problems.

We don't currently have a solution of how to fix that, but I am opening this to track that and close https://github.com/rapidsai/ucx-py/issues/370.

**Comment by pentschev** [comment]
One such approach that was suggested by the Spark team is to use bounce buffers for the UCX transfers, that would likely allow us to use managed memory for other allocations and default allocations for the bounce buffer.

**Comment by jakirkham** [comment]
That sounds like a good idea. Though I thought we weren't able to do this because of UCX capturing allocations? Is there a scenario where that wouldn't be an issue?

**Comment by pentschev** [comment]
The IB registration patch special-cases managed memory and raises an error in those situations: https://github.com/rapidsai/ucx-split-feedstock/blob/5b19834c19607d990cb19da4ee8c0aa1dd15dee1/recipe/ib_registration_cache.patch#L58-L61 . Perhaps we can do some sort of similar workaround for this as well.

**Comment by jakirkham** [comment]
Is it possible UCX already does some copying like this for us?

https://github.com/openucx/ucx/blob/e98a9fb285b8ced245be3c10b41a806c31a907c1/NEWS#L183

**Comment by pentschev** [comment]
I don't think so. CUDA managed memory may work with UCX, but the limitation is CUDA managed memory with CUDA IPC specifically.

**Comment by pentschev** [comment]
A quick update here: it seems like 3-stage pipeline support for managed memory with CUDA IPC is expected to land in UCX inn late August/early September and hopefully make the cut for the next release in November. With that said, we may not need to implement bounce buffers in UCX-Py.

**Comment by jakirkham** [comment]
So in this pipeline they would allocate CUDA memory for us and copy into it. Is that right?

If so, how would the memory allocation occur? Would we have any way to influence how the memory allocation occurs (like ensuring RMM is used)?

**Comment by pentschev** [comment]
> So in this pipeline they would allocate CUDA memory for us and copy into it. Is that right?

No, as far as I understand it, it would only include a bounce buffer type of mechanism internal to UCX to allow CUDA IPC to transfer over the bounce buffers -- which are _not_ managed buffers -- and the application would still be in charge of all the memory allocations that it needs, allowing it to use both managed and unmanaged memory.

**Comment by jakirkham** [comment]
Ok so we'd be responsible for allocating an unmanaged memory that we would hand over to UCX then? Sounds like we are still working with a bounce buffer effectively, but the usage sounds a bit more complex than if we do it ourselves. I'm curious to see how UCX devs think this would work. Is there some rough sketch of this somewhere?

**Comment by pentschev** [comment]
> Ok so we'd be responsible for allocating an unmanaged memory buffer that we would hand over to UCX then?

No, this would be taken care by UCX internally. The UCX-Py user would still manage the application's memory just as we do today.

> Is there some rough sketch of this somewhere?

I don't think there is, but if anyone has one it would be @Akshay-Venkatesh . Would you mind answer that, Akshay?

**Comment by Akshay-Venkatesh** [comment]
It's pretty much along the lines of the pipeline design discussed here but the goal is to copy application managed memory buffers to pinned device bounce buffers (to be able to leverage cuda-ipc/nvlink) and back.

https://ieeexplore.ieee.org/abstract/document/6270863

**Comment by pentschev** [comment]
Thanks for the clarifications @Akshay-Venkatesh . That's pretty much the same idea we had (probably copied from you) for UCX-Py, but since UCX will soon support that, adding bounce buffers to UCX-Py would be redundant.

**Comment by quasiben** [comment]
Possible resolution with https://github.com/openucx/ucx/pull/5889

**Comment by jakirkham** [comment]
IIUC that PR is treating managed memory the same as other memory on host. So CUDA IPC is not used. Does that sound accurate? Or am I missing something?

**Comment by pentschev** [comment]
> IIUC that PR is treating managed memory the same as other memory on host. So CUDA IPC is not used. Does that sound accurate? Or am I missing something?

That does seem like it, I was trying it locally, and I see around 1GB/s transfers, which is in line with what I normally expect of CUDA transfers going through host via Ethernet. I've asked for clarifications in https://github.com/openucx/ucx/pull/5889#issuecomment-726810807 .

**Comment by pentschev** [comment]
> IIUC that PR is treating managed memory the same as other memory on host. So CUDA IPC is not used. Does that sound accurate? Or am I missing something?

That does seem like it, I was trying it locally, and I see around 1GB/s transfers, which is in line with what I normally expect of CUDA transfers going through host via Ethernet. I've asked for clarifications in https://github.com/openucx/ucx/pull/5889#issuecomment-726810807 .

**Comment by pentschev** [comment]
I got an answer in https://github.com/openucx/ucx/pull/5889#issuecomment-726924368, pasting it below for completeness of this thread as well:

> This PR is to prevent the use of GPUDirectRDMA for managed memory to avoid Data validation issues. This PR enables using staging (through host) protocols for managed memory. we will have staging through CUDA in the near future which uses GPUDirectRDMA and NVLINK.

**Comment by pentschev** [comment]
This is definitely resolved in UCX by the use of CUDA staging buffers `UCX_RNDV_FRAG_MEM_TYPES=cuda`, although it will be enabled by default under the right circumstances via the `UCX_RNDV_FRAG_MEM_TYPES=host,cuda` default. Closing.
